cmd/flagger/main.go

```diff
@@ -135,6 +138,14 @@ func main() {
 		logger.Fatalf("Error building mesh clientset: %v", err)
 	}
 
+	var consulClient *consulapi.Client
+	if enableConsulClient {
+		consulClient, err = consulapi.NewClient(consulapi.DefaultConfig())
+		if err != nil {
+			logger.Fatalf("Error building consul client: %v", err)
+		}
+	}
```

Comment:
IMO, we'd better initialize any mesh provider client from`clientset.NewForConfig` or `pkg/router` in order to abstract mesh provider implementations in this main function.


How about :
1. define `NewConsulConnectRouter`  in `pkg/router/connect.go` and initialize the consul client inside the function.
2. call `NewConsulConnectRouter` in `Faction.MeshRouter` https://github.com/weaveworks/flagger/blob/master/pkg/router/factory.go#L55

and in order to prevent multiple instantiations of consul client, we could use `sync.Once` function.

Reply:
Hi, thanks for the review!

Out of my experience, wiring and configuration of applications should happen near the main function. And there is no much point in abstracting out things in main, as main anyway knows all concrete details about the whole application.

I am okay to change this, but just want to compare the pros/cons. Can you explain more why we should move this initialization? Is it because you want to make the initialization of the Consul client happen only if the `connect` provider is chosen?

Reply:
> Is it because you want to make the initialization of the Consul client happen only if the connect provider is chosen?

yes this is exactly what i have in my mind, and also it makes it possible for us to remove `enableConsulClient` variable.

what you said is absolutely right, so if there's a better workaround, i want to find the one